### PR TITLE
Refactors foam code with some of TG foam code

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
Refactors foam code, also making them apply reagents to mob even if metal foam.
also gets rid of stinky spawn()

## Why It's Good For The Game
Refactor good

## Changelog
:cl:
refactor: refactors foam code.
/:cl:

### CODE REVIEW

**Comment by TiviPlus** [comment]
Didnt even compile 💢

**Comment by SplinterGP** [comment]
> Didnt even compile 💢

i know right i forgot something

File: code/game/objects/effects/effect_system/foam.dm [hunk 1]
```diff
@@ -19,82 +19,103 @@
...
+		return
+	spread_foam()
+
+///Applies foam reagents reaction on the mob OR anything overrided by some foam type.
+/obj/effect/particle_effect/foam/proc/foam_mob(mob/living/L)
+	if(lifetime < 1)
+		return FALSE
+	if(!isliving(L))
+		return FALSE
+	var/fraction = 1/initial(reagent_divisor)
```
Comment (TiviPlus) [inline code/game/objects/effects/effect_system/foam.dm:88]: why all these initials on reagent division

**Review by TiviPlus — CHANGES_REQUESTED** [review]
.

File: code/game/objects/effects/effect_system/foam.dm [hunk 2]
```diff
@@ -19,82 +19,103 @@
...
 /obj/effect/particle_effect/foam/process()
-	if(--amount < 0)
+	lifetime--
+	if(lifetime < 1)
+		kill_foam()
 		return
 
-
+	var/fraction = 1/initial(reagent_divisor)
+	for(var/obj/O in range(0, src))
```
[the remark this answers is not in the document]
Comment (SplinterGP) [reply code/game/objects/effects/effect_system/foam.dm:65]: TG uses this, might replace it for something else.

[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/game/objects/effects/effect_system/foam.dm:65]: Yea its awful replace it

File: code/game/objects/effects/effect_system/foam.dm [hunk 3]
```diff
@@ -19,82 +19,103 @@
...
+		if(O.type == src.type)
+			continue
+		if(lifetime % reagent_divisor)
+			reagents.reaction(O, VAPOR, fraction)
+	var/hit = 0
+	for(var/mob/living/L in range(0, src))
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/game/objects/effects/effect_system/foam.dm:71]: same

File: code/game/objects/effects/effect_system/foam.dm (hunk 1 above)
Comment (TiviPlus) [reply code/game/objects/effects/effect_system/foam.dm:88]: standing

File: code/game/objects/effects/effect_system/foam.dm [hunk 4]
```diff
@@ -19,82 +19,103 @@
...
-		if(F)
+		var/obj/effect/particle_effect/foam/oldF = locate() in T
+		if(oldF)
 			continue
-
-		F = new(T, metal)
+		
+		for(var/mob/living/L in T)
+			foam_mob(L)
+		var/obj/effect/particle_effect/foam/F = new src.type(T)
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/game/objects/effects/effect_system/foam.dm:108]: still standing, remove the src from src.type

File: code/game/objects/effects/effect_system/foam.dm [hunk 5]
```diff
@@ -112,52 +133,39 @@
...
-		amount = round(sqrt(amt / 3), 1)
-		if(istype(loca, /turf/))
-			location = loca
-		else
-			location = get_turf(loca)
-
-		carried_reagents = list()
+/datum/effect_system/foam_spread/New()
+	..()
+	chemholder = new /obj()
```
Comment (TiviPlus) [inline code/game/objects/effects/effect_system/foam.dm:141]: whyyyyyyyyyyyyyyyyyyyyyy

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by SplinterGP** [comment]
can do tomorrow

**Pull request closed without merging**